src/blocks/packages/apt.rs

```diff
@@ -0,0 +1,133 @@
+                .stdout,
+        )
+        .error("Problem capturing apt-cache command output")?;
+
+        let phased_regex = regex!(r".*\(phased (\d+)%\).*");
+        Ok(match phased_regex.captures(&output) {
+            Some(matches) => &matches[1] != "100",
+            None => false,
+        })
+    }
+
+    async fn setup(&mut self) -> Result<()> {
+        let mut cache_dir = env::temp_dir();
+        cache_dir.push("i3rs-apt");
+        if !cache_dir.exists() {
+            create_dir_all(&cache_dir)
+                .await
+                .error("Failed to create temp dir")?;
+        }
+
+        let apt_config = format!(
+            "Dir::State \"{}\";\n
+         Dir::State::lists \"lists\";\n
+         Dir::Cache \"{}\";\n
+         Dir::Cache::srcpkgcache \"srcpkgcache.bin\";\n
+         Dir::Cache::pkgcache \"pkgcache.bin\";",
+            cache_dir.display(),
+            cache_dir.display(),
+        );
+
+        let mut config_file = cache_dir;
+        config_file.push("apt.conf");
+        let config_file = config_file.to_str().unwrap();
+
+        self.config_file = config_file.to_string();
+
+        let mut file = File::create(&config_file)
+            .await
+            .error("Failed to create config file")?;
+        file.write_all(apt_config.as_bytes())
+            .await
+            .error("Failed to write to config file")?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Backend for Apt {
+    fn name(&self) -> &str {
+        "apt"
+    }
+
+    async fn get_updates_list(&self) -> Result<Vec<String>> {
+        Command::new("apt")
+            .env("APT_CONFIG", &self.config_file)
+            .args(["update"])
+            .stdout(Stdio::null())
+            .stdin(Stdio::null())
+            .spawn()
+            .error("Failed to run `apt update`")?
+            .wait()
+            .await
+            .error("Failed to run `apt update`")?;
+        let stdout = Command::new("apt")
+            .env("LANG", "C")
+            .env("APT_CONFIG", &self.config_file)
+            .args(["list", "--upgradable"])
+            .output()
+            .await
+            .error("Problem running apt command")?
+            .stdout;
+
+        let updates = String::from_utf8(stdout).error("apt produced non-UTF8 output")?;
+        let updates: Vec<String> = updates
+            .lines()
+            .filter(|line| line.contains("[upgradable"))
+            .filter_map(|update_line| {
+                let is_phased_update =
+                    async { self.is_phased_update(update_line).await.unwrap_or(false) };
+
+                Some(update_line.to_string()).filter(|_| {
+                    !self.ignore_phased_updates
+                        || !tokio::task::block_in_place(|| {
+                            tokio::runtime::Runtime::new()
+                                .unwrap()
+                                .block_on(is_phased_update)
+                        })
```

Comment:
Spawning a new tokio runtime for each update is not great. Since you can't use `.await` in a closure, you should just use a regular for loop and `.push()` updates into the vector.

Reply:
I predicted you may say this :smile:

Reply:
Done